src/crypto/secretstream/secretstream_macros.rs

```diff
@@ -0,0 +1,353 @@
+/// not the end of the stream.
+/// For example, a huge JSON string sent as multiple chunks can use this tag to
+/// indicate to the application that the string is complete and that it can be
+/// decoded. But the stream itself is not closed, and more data may follow.
+const TAG_PUSH: u8 = $tag_push as u8;
+
+/// Tag rekey: "forget" the key used to encrypt this message and the previous
+/// ones, and derive a new secret key.
+const TAG_REKEY: u8 = $tag_rekey as u8;
+
+/// Tag final: indicates that the message marks the end of the stream and erases
+/// the secret key used to encrypt the previous sequence.
+const TAG_FINAL: u8 = $tag_final as u8;
+
+/// A tag is encrypted and attached to each message before the authentication
+/// code is generated over all data. A typical encrypted stream simply attaches
+/// `0` as a tag to all messages, except the last one which is tagged as
+/// `Tag::Final`. When decrypting the tag is retrieved and may be used.
+#[derive(Debug, PartialEq)]
+pub enum Tag {
+    /// Message, the most common tag, that doesn't add any information about the
+    /// nature of the message.
+    Message,
+    /// Push: indicates that the message marks the end of a set of messages, but
+    /// not the end of the stream.
+    /// For example, a huge JSON string sent as multiple chunks can use this tag
+    /// to indicate to the application that the string is complete and that it
+    /// can be decoded. But the stream itself is not closed, and more data may
+    /// follow.
+    Push,
+    /// Rekey: "forget" the key used to encrypt this message and the previous
+    /// ones, and derive a new secret key.
+    Rekey,
+    /// Final: indicates that the message marks the end of the stream and erases
+    /// the secret key used to encrypt the previous sequence.
+    Final,
+}
+
+impl Tag {
+    /// Returns the corresponding `Tag` given a `u8`, else `Err(())`.
+    fn from_u8(tag: u8) -> Result<Tag, ()> {
+        match tag {
+            TAG_MESSAGE => Ok(Tag::Message),
+            TAG_PUSH => Ok(Tag::Push),
+            TAG_REKEY => Ok(Tag::Rekey),
+            TAG_FINAL => Ok(Tag::Final),
+            _ => Err(())
+        }
+    }
+}
+
+new_type! {
+    /// `Key` for symmetric authenticated encryption.
+    ///
+    /// When a `Key` goes out of scope its contents will be overwritten in
+    /// memory.
+    secret Key(KEYBYTES);
+}
+
+new_type! {
+    /// An encrypted stream starts with a short header, whose size is HEADERBYTES bytes.
+    /// That header must be sent/stored before the sequence of encrypted messages,
+    /// as it is required to decrypt the stream.
+    public Header(HEADERBYTES);
+}
+
+/// `gen_key()` randomly generates a secret key
+///
+/// THREAD SAFETY: `gen_key()` is thread-safe provided that you have
+/// called `sodiumoxide::init()` once before using any other function
+/// from sodiumoxide.
+pub fn gen_key() -> Key {
+    let mut key = [0; KEYBYTES];
+    randombytes_into(&mut key);
```

Comment:
Does it really have to be randombytes?

Reply:
What do you mean exactly? You mean to use a different method? I used here what was used in other modules to generate a key.

Reply:
OK.

Reply:
AFAIK that is totally fine (it's what's also done in libsodium, see [here](https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_secretstream/xchacha20poly1305/secretstream_xchacha20poly1305.c#L36). The [docs](https://libsodium.gitbook.io/doc/secret-key_cryptography/secretstream#encryption) also explain that any key can be used (the function referred to is `crypto_secretstream_xchacha20poly1305_keygen`)

> Note that using this function is not required to obtain a suitable key: the secretstream API can use any secret key whose size is crypto_secretstream_xchacha20poly1305_KEYBYTES bytes.
> Network protocols can leverage the key exchange API in order to get a shared key that can be used to encrypt streams. Similarly, file encryption applications can use the password hashing API to get a key that can be used with the functions below.

Reply:
```
void
crypto_secretstream_xchacha20poly1305_keygen
   (unsigned char k[crypto_secretstream_xchacha20poly1305_KEYBYTES])
{
    randombytes_buf(k, crypto_secretstream_xchacha20poly1305_KEYBYTES);
}
```

So there is a special function to generate a key even though it equals to randombytes_into. Assume there will be another module and there will be another `crypto_secretstream_MODULE_keygen` with a different strategy. Is it possible?

Reply:
Yes, I believe every module has its own `keygen()`, one example of this is in the `auth` module: https://github.com/sodiumoxide/sodiumoxide/blob/master/src/crypto/auth/auth_macros.rs#L36

Reply:
If you use stream_module! on a module with keygen!=randombytes, the logic will be incorrect. If there is a stream module with such property, we should add crypto_secretstream_xchacha20poly1305_keygen as a $variable in macros. If there is no such a module, we shouldn't care.

Reply:
As far as I know, all `keygen()` functions are always random bytes (except for keypair's of course). I guess they are redefined in every module for consistency reasons and the fact that the key sizes can differ.

Reply:
Ok, resolved